### DISCUSSION TITLE
Benchmark Singlestore with hammerDB

### DISCUSSION THREAD

**Discussion by dhgokul** [TPC-Council/HammerDB#656] [Q&A] [answered]
Can you please provide details on whether load testing can be performed using HammerDB for singlestore ?
https://www.singlestore.com/

I configured hammercli in Redhat OS. Installed the SQL client  executed the 'librarycheck', and confirmed the ODBC connector loaded successfully for MySQL. However, when I created a tcl file with the engine set as SIngleStore, I encountered an issue. PFA

![Screenshot 2024-01-15 at 19 17 20](https://github.com/TPC-Council/HammerDB/assets/5689804/62b22abc-c008-49aa-823b-5b97229a0293)
![Screenshot 2024-01-15 at 19 20 24](https://github.com/TPC-Council/HammerDB/assets/5689804/a55e9d0c-053f-4db3-874f-e802dcac7034)

**Comment by sm-shaw** [comment, marked as answer]
We have not tested singlestore, however if it is fully MySQL compatible then it will work with the HammerDB MySQL option. 

However, you are mixing up Microsoft SQL Server and MySQL. In the screenshot you highlight the SQL Server driver not the MySQL one. It is the MySQL one you need. Install a MySQL client library, or add an existing one to your path. 

Follow the guidance here https://www.hammerdb.com/docs/ch01s10.html - when librarycheck shows the MySQL library loads then use the MySQL build and test against your MySQL environment.

**Reply by dhgokul** [reply]
@sm-shaw  Thanks for your response, I have configured the  MySQL client and still having connection issue.

![Screenshot 2024-01-16 at 13 16 58](https://github.com/TPC-Council/HammerDB/assets/5689804/1214a2ec-dc2f-4c39-bfb4-2d0a44a35868)

**SingleStore Overview:**
SingleStore is designed to be a highly performant and scalable database, and to achieve this, it has made some changes to the MySQL architecture. Some of the key differences between SingleStore and MySQL include:

SingleStore uses a distributed architecture, while MySQL uses a centralized architecture.
SingleStore can handle a larger number of concurrent queries and has a faster query performance than MySQL.
SingleStore has a specialized storage engine that is optimized for high-performance analytical workloads.
SingleStore includes additional features such as real-time analytics, geospatial data support, and a unique "Memorystore" option that allows for faster data retrieval and faster analytics.
SingleStore has a MySQL compatibility layer that allows it to support a subset of the MySQL SQL syntax and functions. However, some MySQL features, such as certain types of stored procedures, triggers and views, are not currently supported by SingleStore.

**Reply by sm-shaw** [reply]
To be clear, HammerDB supports MySQL and MariaDB and does not support singlestore or other derived versions. 

You show that the library has loaded, so if the database is MySQL compatible, then HammerDB should be able to connect to it. Further steps are for you to resolve with connecting to singlestore. 

When you say "certain types of stored procedures" if this means singlestore does not support stored procedures then there is an option to use client SQL for MariaDB and MySQL https://github.com/TPC-Council/HammerDB/pull/602 however note that not using stored procedures is always going to slower than when you can use them as described here https://www.hammerdb.com/blog/uncategorized/why-you-should-benchmark-your-database-using-stored-procedures/
